File: 2021/07/29/pascal_triangle.py
"""
https://leetcode.com/problems/pascals-triangle/
Author: Andrew Jarombek
Date: 7/28/2021
"""

from typing import List


def generate(num_rows: int) -> List[List[int]]:
    result = [[1]]

    for row in range(1, num_rows):
        row_result = [1]

        for i in range(1, row):
            row_result.append(result[-1][i - 1] + result[-1][i])

        row_result.append(1)
        result.append(row_result)

    return result


if __name__ == '__main__':
    result = generate(5)
    assert len(result) == 5
    assert result[0] == [1]
    assert result[1] == [1, 1]
    assert result[2] == [1, 2, 1]
    assert result[3] == [1, 3, 3, 1]
    assert result[4] == [1, 4, 6, 4, 1]

File: 2021/07/29/PascalTriangle.java
import java.util.ArrayList;
import java.util.List;

/**
 * https://leetcode.com/problems/pascals-triangle-ii
 * @author Andrew Jarombek
 * @since 7/29/2021
 */

public class PascalTriangle {
    public static List<Integer> getRow(int rowIndex) {
        List<Integer> result = new ArrayList<>();
        result.add(1);

        for (int i = 1; i <= rowIndex; i++) {
            List<Integer> nextResult = new ArrayList<>();
            nextResult.add(1);

            for (int j = 1; j < i; j++) {
                nextResult.add(result.get(j) + result.get(j-1));
            }

            nextResult.add(1);
            result = nextResult;
        }

        return result;
    }

    public static void main(String[] args) {
        List<Integer> result = PascalTriangle.getRow(3);
        assert result.size() == 4;
        assert result.get(0) == 1;
        assert result.get(1) == 3;
        assert result.get(2) == 3;
        assert result.get(3) == 1;
    }
}
